txp/Pos/Txp/Toil/Monad.hs

```diff
@@ -0,0 +1,274 @@
+     = ReaderT GlobalToilEnv (StateT GlobalToilState GlobalToilMBase)
+
+-- | Run given action in some monad capable of getting stakeholders'
+-- stakes and logging.
+runGlobalToilMBase ::
+       forall m a. (WithLogger m)
+    => (StakeholderId -> m (Maybe Coin))
+    -> GlobalToilMBase a
+    -> m a
+runGlobalToilMBase stakeGetter = launchNamedPureLog foldF'
+  where
+    foldF' :: forall x. F StakesLookupF x -> m x
+    foldF' =
+        foldF $ \case
+            StakesLookupF sId f -> f <$> stakeGetter sId
+
+-- | Run 'GlobalToilM' action in some monad capable of getting
+-- stakeholders' stakes and logging.
+runGlobalToilM ::
+       forall m a. (WithLogger m)
+    => GlobalToilEnv
+    -> GlobalToilState
+    -> (StakeholderId -> m (Maybe Coin))
+    -> GlobalToilM a
+    -> m (a, GlobalToilState)
+runGlobalToilM env gts stakeGetter =
+    runGlobalToilMBase stakeGetter . usingStateT gts . usingReaderT env
+
+-- | Get stake of a given stakeholder.
+getStake :: StakeholderId -> GlobalToilM (Maybe Coin)
+getStake id =
+    (<|>) <$> use (gtsStakesView . svStakes . at id) <*> baseLookup id
+  where
+    baseLookup :: StakeholderId -> GlobalToilM (Maybe Coin)
+    baseLookup i =
+        lift $ lift $ lift $ F $ \kPure kFree -> kFree (StakesLookupF i kPure)
+
+-- | Get total stake of all stakeholders.
+getTotalStake :: GlobalToilM Coin
+getTotalStake =
+    maybe (view gteTotalStake) pure =<< use (gtsStakesView . svTotal)
+
+-- | Set stake of a given stakeholder.
+setStake :: StakeholderId -> Coin -> GlobalToilM ()
+setStake id c = gtsStakesView . svStakes . at id .= Just c
+
+-- | Set total stake of all stakeholders.
+setTotalStake :: Coin -> GlobalToilM ()
+setTotalStake c = gtsStakesView . svTotal .= Just c
+
+-- | Extended version of 'GlobalToilM'. It allows to put extra data
+-- into reader context and extra state. It's needed for explorer which
+-- has more complicated transaction processing.
+type ExtendedGlobalToilM extraEnv extraState =
+    ReaderT (GlobalToilEnv, extraEnv) (
+        StateT (GlobalToilState, extraState) (
+            GlobalToilMBase
+    ))
+
+-- | Natural transformation from 'GlobalToilM to 'ExtendedGlobalToilM'.
+natGlobalToilM :: GlobalToilM ~> ExtendedGlobalToilM __ ___
+natGlobalToilM = zoom _1 . magnify _1
+
+----------------------------------------------------------------------------
+-- Conversions
+----------------------------------------------------------------------------
+
+-- | Lift 'UtxoM' action to 'LocalToilM'.
+utxoMToLocalToilM :: UtxoM ~> LocalToilM
+utxoMToLocalToilM = mapReaderT f
+  where
+    f :: forall a. State UtxoModifier a -> State LocalToilState a
+    f = zoom ltsUtxoModifier
+
+-- | Lift 'UtxoM' action to 'GlobalToilM'.
+utxoMToGlobalToilM :: UtxoM ~> GlobalToilM
+utxoMToGlobalToilM = mapReaderT f . magnify gteUtxo
+  where
+    f :: State UtxoModifier
+      ~> StateT GlobalToilState (NamedPureLogger (F StakesLookupF))
+    f = state . runState . zoom gtsUtxoModifier
```

Comment:
`state . runState` is incidental here, what we care about is generalizing `Identity` to `NamePureLogger (F StakesLookupF)`.

I suggest:

```
utxoMToGlobalToilM = generalizeBase . zoom gtsUtxoModifier . magnify gteUtxo
  where
    generalizeBase = hoist (hoist generalize)
```

and leave `generalizeBase` without a type signature, it only creates noise.

Reply:
Actually I just copied the definition of [zoom'](https://github.com/ekmett/lens/issues/580). This function is present in `serokell-util`, but unfortunately there is only a version for lazy `State`, while I need one for strict `State`.

Reply:
I'm not sure, why is it needed here?

Reply:
If `zoom'` was in `lens` would you object to using it here? Or if there was a version for strict `State` in `serokell-util`?